File: problems/basic/24/solution2.cpp
// Accepted: 4ms (41.68%), 7.78MB (93.40%)

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if (head == nullptr || head->next == nullptr) {
            return head;
        }

        ListNode *left, *right;
        left = head;
        right = head->next;

        left->next = swapPairs(right->next);
        right->next = left;

        return right; // as a new head
    }
};

File: problems/basic/24/solution2.py
# Accepted: 41ms (44.87%), 17.25MB (21.85%)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head

        left, right = head, head.next
        left.next = self.swapPairs(right.next)
        right.next = left

        return right
